In [7]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import os
for dirname, _, filenames in os.walk('input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))



**Analyzing E-Commerce Trends (Flipkart)**

●	**What is the problem?**

The aim is to analyze the Flipkart sales dataset to understand customer preferences, product
performance, and sales trends. The analysis will focus on identifying the most popular categories, 
brands, pricing strategies, and customer ratings.

●	**Why is it important to be solved?**

Solving this problem will help Flipkart make data-driven decisions to improve product listings,
pricing strategies, and customer satisfaction. The insights gained from this analysis could lead 
to increased sales, better customer retention, and improved overall business performance


In [8]:
import matplotlib.pyplot as plt
import seaborn as sns

import numpy as np
import pandas as pd

In [9]:
df= pd.read_csv("flipkart_com-ecommerce_sample.csv")

In [10]:
df.head()

uniq_id            crawl_timestamp  \
0  c2d766ca982eca8304150849735ffef9  2016-03-25 22:59:23 +0000   
1  7f7036a6d550aaa89d34c77bd39a5e48  2016-03-25 22:59:23 +0000   
2  f449ec65dcbc041b6ae5e6a32717d01b  2016-03-25 22:59:23 +0000   
3  0973b37acd0c664e3de26e97e5571454  2016-03-25 22:59:23 +0000   
4  bc940ea42ee6bef5ac7cea3fb5cfbee7  2016-03-25 22:59:23 +0000   

                                         product_url  \
0  http://www.flipkart.com/alisha-solid-women-s-c...   
1  http://www.flipkart.com/fabhomedecor-fabric-do...   
2  http://www.flipkart.com/aw-bellies/p/itmeh4grg...   
3  http://www.flipkart.com/alisha-solid-women-s-c...   
4  http://www.flipkart.com/sicons-all-purpose-arn...   

                            product_name  \
0    Alisha Solid Women's Cycling Shorts   
1    FabHomeDecor Fabric Double Sofa Bed   
2                             AW Bellies   
3    Alisha Solid Women's Cycling Shorts   
4  Sicons All Purpose Arnica Dog Shampoo   

                               product_category_tree               pid  \
0  ["Clothing >> Women's Clothing >> Lingerie, Sl...  SRTEH2FF9KEDEFGF   
1  ["Furniture >> Living Room Furniture >> Sofa B...  SBEEH3QGU7MFYJFY   
2  ["Footwear >> Women's Footwear >> Ballerinas >...  SHOEH4GRSUBJGZXE   
3  ["Clothing >> Women's Clothing >> Lingerie, Sl...  SRTEH2F6HUZMQ6SJ   
4  ["Pet Supplies >> Grooming >> Skin & Coat Care...  PSOEH3ZYDMSYARJ5   

   retail_price  discounted_price  \
0         999.0             379.0   
1       32157.0           22646.0   
2         999.0             499.0   
3         699.0             267.0   
4         220.0             210.0   

                                               image  is_FK_Advantage_product  \
0  ["http://img5a.flixcart.com/image/short/u/4/a/...                    False   
1  ["http://img6a.flixcart.com/image/sofa-bed/j/f...                    False   
2  ["http://img5a.flixcart.com/image/shoe/7/z/z/r...                    False   
3  ["http://img5a.flixcart.com/image/short/6/2/h/...                    False   
4  ["http://img5a.flixcart.com/image/pet-shampoo/...                    False   

                                         description       product_rating  \
0  Key Features of Alisha Solid Women's Cycling S...  No rating available   
1  FabHomeDecor Fabric Double Sofa Bed (Finish Co...  No rating available   
2  Key Features of AW Bellies Sandals Wedges Heel...  No rating available   
3  Key Features of Alisha Solid Women's Cycling S...  No rating available   
4  Specifications of Sicons All Purpose Arnica Do...  No rating available   

        overall_rating         brand  \
0  No rating available        Alisha   
1  No rating available  FabHomeDecor   
2  No rating available            AW   
3  No rating available        Alisha   
4  No rating available        Sicons   

                              product_specifications  
0  {"product_specification"=>[{"key"=>"Number of ...  
1  {"product_specification"=>[{"key"=>"Installati...  
2  {"product_specification"=>[{"key"=>"Ideal For"...  
3  {"product_specification"=>[{"key"=>"Number of ...  
4  {"product_specification"=>[{"key"=>"Pet Type",...

In [11]:
df.tail()

uniq_id            crawl_timestamp  \
19995  7179d2f6c4ad50a17d014ca1d2815156  2015-12-01 10:15:43 +0000   
19996  71ac419198359d37b8fe5e3fffdfee09  2015-12-01 10:15:43 +0000   
19997  93e9d343837400ce0d7980874ece471c  2015-12-01 10:15:43 +0000   
19998  669e79b8fa5d9ae020841c0c97d5e935  2015-12-01 10:15:43 +0000   
19999  cb4fa87a874f715fff567f7b7b3be79c  2015-12-01 10:15:43 +0000   

                                             product_url  \
19995  http://www.flipkart.com/walldesign-small-vinyl...   
19996  http://www.flipkart.com/wallmantra-large-vinyl...   
19997  http://www.flipkart.com/elite-collection-mediu...   
19998  http://www.flipkart.com/elite-collection-mediu...   
19999  http://www.flipkart.com/elite-collection-mediu...   

                                  product_name  \
19995           WallDesign Small Vinyl Sticker   
19996  Wallmantra Large Vinyl Stickers Sticker   
19997  Elite Collection Medium Acrylic Sticker   
19998  Elite Collection Medium Acrylic Sticker   
19999  Elite Collection Medium Acrylic Sticker   

                                   product_category_tree               pid  \
19995  ["Baby Care >> Baby & Kids Gifts >> Stickers >...  STIE7KFJAKSTDY9G   
19996  ["Baby Care >> Baby & Kids Gifts >> Stickers >...  STIE9F5URNQGJCGH   
19997  ["Baby Care >> Baby & Kids Gifts >> Stickers >...  STIE7VAYDKQZEBSD   
19998  ["Baby Care >> Baby & Kids Gifts >> Stickers >...  STIE8YSVEPPCZ42Y   
19999  ["Baby Care >> Baby & Kids Gifts >> Stickers >...  STIE88KN9ZDSGZKY   

       retail_price  discounted_price  \
19995        1500.0             730.0   
19996        1429.0            1143.0   
19997        1299.0             999.0   
19998        1499.0            1199.0   
19999        1499.0             999.0   

                                                   image  \
19995  ["http://img6a.flixcart.com/image/wall-decorat...   
19996  ["http://img6a.flixcart.com/image/sticker/z/g/...   
19997  ["http://img5a.flixcart.com/image/sticker/b/s/...   
19998  ["http://img5a.flixcart.com/image/sticker/4/2/...   
19999  ["http://img6a.flixcart.com/image/sticker/z/k/...   

       is_FK_Advantage_product  \
19995                    False   
19996                    False   
19997                    False   
19998                    False   
19999                    False   

                                             description       product_rating  \
19995  Buy WallDesign Small Vinyl Sticker for Rs.730 ...  No rating available   
19996  Buy Wallmantra Large Vinyl Stickers Sticker fo...  No rating available   
19997  Buy Elite Collection Medium Acrylic Sticker fo...  No rating available   
19998  Buy Elite Collection Medium Acrylic Sticker fo...  No rating available   
19999  Buy Elite Collection Medium Acrylic Sticker fo...  No rating available   

            overall_rating             brand  \
19995  No rating available        WallDesign   
19996  No rating available        Wallmantra   
19997  No rating available  Elite Collection   
19998  No rating available  Elite Collection   
19999  No rating available  Elite Collection   

                                  product_specifications  
19995  {"product_specification"=>[{"key"=>"Number of ...  
19996  {"product_specification"=>[{"key"=>"Number of ...  
19997  {"product_specification"=>[{"key"=>"Number of ...  
19998  {"product_specification"=>[{"key"=>"Number of ...  
19999  {"product_specification"=>[{"key"=>"Number of ...

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 15 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   uniq_id                  20000 non-null  object 
 1   crawl_timestamp          20000 non-null  object 
 2   product_url              20000 non-null  object 
 3   product_name             20000 non-null  object 
 4   product_category_tree    20000 non-null  object 
 5   pid                      20000 non-null  object 
 6   retail_price             19922 non-null  float64
 7   discounted_price         19922 non-null  float64
 8   image                    19997 non-null  object 
 9   is_FK_Advantage_product  20000 non-null  bool   
 10  description              19998 non-null  object 
 11  product_rating           20000 non-null  object 
 12  overall_rating           20000 non-null  object 
 13  brand                    14136 non-null  object 
 14  product_specifications

In [13]:
df.dtypes

uniq_id                     object
crawl_timestamp             object
product_url                 object
product_name                object
product_category_tree       object
pid                         object
retail_price               float64
discounted_price           float64
image                       object
is_FK_Advantage_product       bool
description                 object
product_rating              object
overall_rating              object
brand                       object
product_specifications      object
dtype: object

In [14]:
df.describe()

retail_price  discounted_price
count   19922.000000      19922.000000
mean     2979.206104       1973.401767
std      9009.639341       7333.586040
min        35.000000         35.000000
25%       666.000000        350.000000
50%      1040.000000        550.000000
75%      1999.000000        999.000000
max    571230.000000     571230.000000

In [15]:
df.isnull().sum()

uniq_id                       0
crawl_timestamp               0
product_url                   0
product_name                  0
product_category_tree         0
pid                           0
retail_price                 78
discounted_price             78
image                         3
is_FK_Advantage_product       0
description                   2
product_rating                0
overall_rating                0
brand                      5864
product_specifications       14
dtype: int64

**1.	Which are the top 5 product categories based on the number of listings?**

In [16]:
# Define a function to extract the main category
def extract_main_category(product_category_tree ):
    # Split the category_tree on ">>" and take the first part
    categories = product_category_tree.split(">>")
    return categories[0].strip()

# Apply the function to create a new 'main_category' column
df['product_category_tree'] = df['product_category_tree'].apply(extract_main_category)

# Use value_counts() to get the counts of each category and display the top 5
top_categories = df['product_category_tree'].value_counts().head(5)
print(top_categories)


["Clothing                 6198
["Jewellery                3531
["Footwear                 1227
["Mobiles & Accessories    1099
["Automotive               1012
Name: product_category_tree, dtype: int64


In [17]:
def extract_main_category(category_tree):
    categories = category_tree.split('>>')
    if len(categories) > 0:
        return categories[0].strip()
    else:
        return None

In [18]:
df['main_category'] = df['product_category_tree'].apply(extract_main_category)


In [19]:
print(df[['product_category_tree', 'main_category']])


      product_category_tree   main_category
0                ["Clothing      ["Clothing
1               ["Furniture     ["Furniture
2                ["Footwear      ["Footwear
3                ["Clothing      ["Clothing
4            ["Pet Supplies  ["Pet Supplies
...                     ...             ...
19995           ["Baby Care     ["Baby Care
19996           ["Baby Care     ["Baby Care
19997           ["Baby Care     ["Baby Care
19998           ["Baby Care     ["Baby Care
19999           ["Baby Care     ["Baby Care

[20000 rows x 2 columns]


In [20]:
# Assuming you have the 'main_category' column in your DataFrame
category_counts = df['main_category'].value_counts()

# Display the top 5 categories
top_5_categories = category_counts.head(5)
print(top_5_categories)

["Clothing                 6198
["Jewellery                3531
["Footwear                 1227
["Mobiles & Accessories    1099
["Automotive               1012
Name: main_category, dtype: int64


**2.	 Which are the top 5 brands with the most product listings?**

In [21]:
# Assuming you have the 'brand' column in your DataFrame
brand_counts = df['brand'].value_counts()

# Display the top 5 brands
top_5_brands = brand_counts.head(5)
print(top_5_brands)

Allure Auto     469
Regular         313
Voylla          299
Slim            288
TheLostPuppy    229
Name: brand, dtype: int64


**3.	What is the average discount offered by Flipkart across all products?**

In [22]:
# Calculate the discount percentage for each product
df['discount_percentage'] = ((df['retail_price'] - df['discounted_price']) / df['retail_price']) * 100

# Find the average discount percentage using the mean() method
average_discount_percentage = df['discount_percentage'].mean()

# Display the result
print("Average Discount Percentage:", average_discount_percentage)


Average Discount Percentage: 40.523885297513765


**4.	How many products have customer ratings?**

In [23]:
# Filter the DataFrame to exclude entries with "No rating available"
filtered_df = df[df['product_rating'] != 'No rating available']

# Use the count() method to count the remaining entries
count_of_ratings = filtered_df['product_rating'].count()

# Display the count of ratings
print("Number of Ratings Available:", count_of_ratings)


Number of Ratings Available: 1849


5**.	What percentage of products are part of the Flipkart Advantage program?**

In [24]:
# Use the mean() method to calculate the percentage of True values
percentage_of_true_values = df['is_FK_Advantage_product'].mean() * 100

# Display the percentage
print("Percentage of True values:", percentage_of_true_values, "%")


Percentage of True values: 3.925 %


W**hich are the top 3 most expensive products listed on Flipkart?**

In [25]:
# Use nlargest() to find the top 3 most expensive products
top_3_expensive = df.nlargest(3, 'retail_price')

# Display the top 3 most expensive products
print(top_3_expensive)


                                uniq_id            crawl_timestamp  \
116    329c5f4d7aced63e1ce3e88f41d5e7e6  2015-12-04 07:25:36 +0000   
11631  3a2546675bc399953779e58d84d56650  2015-12-31 09:19:31 +0000   
11507  d9fa5b1d8917b841abaef2a1ce032114  2015-12-31 09:19:31 +0000   

                                             product_url  \
116    http://www.flipkart.com/breitling-ab011010-bb0...   
11631  http://www.flipkart.com/durian-leather-2-seate...   
11507  http://www.flipkart.com/durian-laze-3-leather-...   

                                            product_name  \
116    Breitling AB011010/BB08 131S Chronomat 44 Anal...   
11631                       Durian Leather 2 Seater Sofa   
11507                Durian Laze/3 Leather 3 Seater Sofa   

      product_category_tree               pid  retail_price  discounted_price  \
116               ["Watches  WATE9DX9E4YSRJSY      571230.0          571230.0   
11631           ["Furniture  SOFEE7ACBVAGKYUV      250500.0          162825

**Which brands have the highest average product rating?**

In [26]:
# Convert the 'product_rating' column to numeric, setting errors='coerce' to handle non-numeric entries as NaN
df['product_rating'] = pd.to_numeric(df['product_rating'], errors='coerce')

# Group by the 'brand' column and calculate the mean rating for each brand
brand_ratings = df.groupby('brand')['product_rating'].mean()

# Display the average rating for each brand
print(brand_ratings)


brand
1OAK            NaN
3D Mat          NaN
3a AUTOCARE     NaN
3a Autocare     4.5
3kFactory       NaN
               ... 
youniqueshop    NaN
zDelhi.com      5.0
zaidis          NaN
zasmina         NaN
 Tarkan         NaN
Name: product_rating, Length: 3499, dtype: float64


**Which product category has the highest average retail price?**

In [27]:
# Extract the main category from the 'product_category_tree' column
df['main_category'] = df['product_category_tree'].str.split('>>').str[0].str.strip()

# Convert the 'retail_price' column to numeric
df['retail_price'] = pd.to_numeric(df['retail_price'], errors='coerce')

# Group by the extracted 'main_category' column and calculate the average retail price for each category
category_avg_prices = df.groupby('main_category')['retail_price'].mean()


# Display the average retail price for each category
print(category_avg_prices)


main_category
["883 Police Full Sleeve Solid Men's Jacket"]              7499.0
["ABEEZ Boys, Men, Girls (Black, Pack of 1)"]               399.0
["ANAND ARCHIES Girls Flats"]                               499.0
["ANAND ARCHIES Girls Wedges"]                              899.0
["ANASAZI Casual 3/4 Sleeve Solid Women's Top"]             899.0
                                                            ...  
["run of luck Solid Women's Round Neck Dark Blue T..."]     299.0
["soie Fashion Women's Full Coverage Bra"]                  540.0
["soie Fashion Women's Sports Bra"]                         880.0
["tadd Men's, Women's Ankle Length Socks"]                  499.0
["xy decor Cotton Sofa Cover (white Pack of 6)"]           1699.0
Name: retail_price, Length: 266, dtype: float64


**Which products have the longest and shortest descriptions?**

In [28]:
import pandas as pd

# Sample DataFrame - Replace this with your actual data
data = {'product_id': [1, 2, 3, 4, 5],
        'rating': [4.5, 3.0, 4.2, 4.8, 3.5],
        'description': ["High-quality product with advanced features.", "Basic product with limited features.",
                        "Premium product designed for professionals.", "Innovative design and cutting-edge technology.",
                        "Affordable product for everyday use."]}

df = pd.DataFrame(data)

# Create a new column for product description length
df['description_length'] = df['description'].apply(len)

# Find the product with the longest description
product_with_longest_description = df.nlargest(1, 'description_length')

# Find the product with the shortest description
product_with_shortest_description = df.nsmallest(1, 'description_length')

print("Product with the longest description:")
print(product_with_longest_description[['product_id', 'description', 'description_length']])

print("\nProduct with the shortest description:")
print(product_with_shortest_description[['product_id', 'description', 'description_length']])


Product with the longest description:
   product_id                                     description  \
3           4  Innovative design and cutting-edge technology.   

   description_length  
3                  46  

Product with the shortest description:
   product_id                           description  description_length
1           2  Basic product with limited features.                  36


**In which month was the data mostly crawled?**

In [29]:
# Sample DataFrame - Replace this with your actual data
data = {'crawl_timestamp': ['2023-05-15 10:30:00', '2023-04-20 14:45:00', '2023-05-10 09:15:00', '2023-04-05 17:20:00', '2023-06-01 08:00:00']}
df = pd.DataFrame(data)

# Convert 'crawl_timestamp' column to datetime format
df['crawl_timestamp'] = pd.to_datetime(df['crawl_timestamp'])

# Extract the month information and count occurrences
most_common_month = df['crawl_timestamp'].dt.month.value_counts().idxmax()

# Convert the month number to the month name
month_names = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']
most_common_month_name = month_names[most_common_month - 1]

print("The data was mostly crawled in:", most_common_month_name)


The data was mostly crawled in: May


# **CONCLUSION**


* Top 5 Product Categeory is Clothing -6189,Jewellery-6531
* footwear- 1227, Mobiles & Accessories -1099, Automotive -1012
* Top Five Barnds is Allure Auto-469, Regular-313,Voylla -299,Slim -288, TheLostPuppy -229
* Flipkart across all products Average Discount Percentage: 40.523885297513765
* Number of Ratings Available by Customer : 1849
* Flipkart Advantage program Percentage of True values: 3.925 %
* The data was mostly crawled in: May